### ISSUE TITLE
AWS::Lambda::Function - RDS Proxy configuration

### ISSUE THREAD

**Issue by grauj-aws** [aws-cloudformation/cloudformation-coverage-roadmap#750]
## 2. Scope of request 

- Lambda supports adding an RDS proxy for its functions, as noted [on this documentation](https://docs.aws.amazon.com/lambda/latest/dg/configuration-database.html)
- It's currently not possible to configure this setting using the [AWS::Lambda::Function resource](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html)

## 3. Expected behavior

- Support adding an RDS proxy for Lambda functions created and managed by CloudFormation

## 4. Suggest specific test cases

- Sample of the expected usage:
```
Type: AWS::Lambda::Function
Properties: 
  DatabaseProxy: String
```

## 5. Helpful Links to speed up research and evaluation

- [AWS::Lambda::Function](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html)
- [AWS::RDS::DBProxy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rds-dbproxy.html)
- [Lambda proxy documentation](https://docs.aws.amazon.com/lambda/latest/dg/configuration-database.html)

## 6. Category

1. Compute (EC2, ECS, EKS, Lambda...)

**Comment by crazy54** [comment]
This is important to a CFN template I am writing, please add this to CFN!

**Comment by teknogeek0** [comment]
Hi there,
  The behavior in the Lambda console uses the standard RDS APIs to create Proxy resources for you, but it is not a special behavior of the Lambda APIs themselves. Today you would just create a resource using the AWS::RDS::DBProxy resource and then pass to the Lambda function the db connection information (endpoint, user/pass as required). There is no other special magic that connects to the two.

Thanks,
- Chris Munns - Serverless Developer Advocacy Lead.
munns@amazon.com

**Comment by davidgatti** [comment]
Sorry @teknogeek0, didn't get any of that. Can you rephrase?

**Comment by benbridts** [comment]
@davidgatti 
If you create a Lambda Function with the web interface ("the Lambda console"), the website will use RDS APIs to create an RDS DBProxy for you, just like you can do yourself by using the CLI or the AWS::RDS::DBProxy resource in CloudFormation.

If you look in the linked documentation you see the sentence "When the proxy is available, configure your function to connect to the proxy endpoint instead of the database endpoint.", meaning that you still have to update your function to connect to the Proxy instead of using the Database directly.

---------
Or to put everything together in CloudFormation.

Let's say you have this as how you define your Lambda Function, using the "ENDPOINT" environment variable to store the Database to connect to:

```
Resources:
  DatabaseInstance:
    Type: AWS::RDS::DBInstance
    Properties:
      # ...
  Function:
    Type: AWS::Lambda::Function
    Properties: 
      # ...
      Environment:
        Variables:
          ENDPOINT: !GetAtt DatabaseInstance.Endpoint.Address
```
you can add the Proxy as a separate resource, connect it to the instance, and have CloudFormation create it.
We also update the environment variable to point to the new endpoint (I didn't check to see if the endpoints are in the same format, you might need a code change too)

```
 Resources:
  Proxy:
    Type: AWS::RDS::DBProxy
    Properties:
      # ...
  ProxyTargetGroup:
    Type: AWS::RDS::DBProxyTargetGroup
    Properties:
      # ...
      DBInstanceIdentifiers: 
        - !Ref DatabaseInstance
      DBProxyName: !Ref Proxy

  DatabaseInstance:
    Type: AWS::RDS::DBInstance
    Properties:
      # ...
  Function:
    Type: AWS::Lambda::Function
    Properties: 
      # ...
      Environment:
        Variables:
          ENDPOINT: !GetAtt Proxy.Endpoint
```

**Comment by davidgatti** [comment]
OK I see what is going on. The issue here dose not explain the whole situation. The proposed solution dose not work if you want to connect to the proxy using a IAM Role, thus needing to generate a token and not use a password in your code. For this to happen, you need to attach the proxy to the lambda, and what role and secret manager to use. And only then your code within lambda will be able to make tokens. 

Your solution of course will work, if you want to use a regular password.

I hope this clarifies the issue.

**Comment by benbridts** [comment]
@davidgatti Those settings are similarly not part of the Lambda function. You configure the Authentication on the [DBProxy Resource](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-rds-dbproxy-authformat.html#cfn-rds-dbproxy-authformat-iamauth), and add the right permissions to the Role that's used by the Lambda function.

**Comment by davidgatti** [comment]
Then what is purpose of the Proxy section in the AWS Console for the Lambda?

**Comment by benbridts** [comment]
It's there to make it more convenient if you're not using IaC. If you're using the [cli](https://docs.aws.amazon.com/cli/latest/reference/lambda/create-function.html), there isn't a Proxy section either.

------

As a side note: you could try a tool like [Former2](https://former2.com/) to generate a template based on existing resources to make it easier to get the right configuration:
- create resources using the console
- Use Former2 to scan your account and select all created resources (this must include all the resources that Lambda creates for you, like AWS::IAM::Role, AWS::Lambda::Permission (when using triggers), AWS::RDS::DBProxy, AWS::RDS:: DBProxyTargetGroup, ...)
- Deploy with CloudFormation
- Remove the resources created via the Console

**Comment by davidgatti** [comment]
As always, confusing AWS - So despite telling you make a custom role that allows lambda to this and that, they added a section where they tell you to make all the roles. Anyway, in this case this Issue is useless. 

Thank you @benbridts

**Comment by davidgatti** [comment]
@grauj-aws you can close this issue.
